CodeEdit/NavigatorSidebar/ProjectNavigator/OutlineView/OutlineMenu.swift

```diff
@@ -169,11 +177,45 @@ final class OutlineMenu: NSMenu {
         item?.showInFinder()
     }
 
+    /// Action that opens the item, identical to clicking it.
+    @objc
+    private func openInTab() {
+        if let item = item {
+            workspace?.openTab(item: item)
+        }
+    }
+
+    /// Action that opens in an external editor
+    @objc
+    private func openWithExternalEditor() {
+        item?.openWithExternalEditor()
+    }
+
+    // TODO: allow custom file names
+    /// Action that creates a new untitled file
+    @objc
+    private func newFile() {
+        item?.addFile(fileName: "untitled")
+    }
```

Comment:
If a file already exists it will crash with:

> Thread 1: Fatal error: Duplicate keys of type 'FileItem' were found in a Dictionary.

Reply:
In [CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift](https://github.com/CodeEditApp/CodeEdit/pull/655/files#diff-36567e2ccad007c9513fbeaf771729619fe2ab185cebff08fe8a467a16b1a984) it does check for duplicates? This error should not be happening. Can you check if the addFile function has been changed on your end? It should look like
```swift
public func addFile(fileName: String) {
    let fileUrl = url.appendingPathComponent(fileName)
    // check if folder, if it is create file under self
    var fileUrl = (self.isFolder ?
        self.url.appendingPathComponent(fileName) :
        self.url.deletingLastPathComponent().appendingPathComponent(fileName))

    // if a file/folder with the same name exists, add a number to the end.
    var fileNumber = 0
    while FileItem.fileManger.fileExists(atPath: fileUrl.path) {
        fileNumber += 1
        fileUrl = fileUrl.deletingLastPathComponent().appendingPathComponent("\(fileName)\(fileNumber)")
    }

    // create the file
    FileItem.fileManger.createFile(
        atPath: fileUrl.path,
        contents: nil,
        attributes: [FileAttributeKey.creationDate: Date()]
    )
}
```

Reply:
looks the same. Pulled it straight from your branch.

Reply:
This is actually an error with the new editor. If you close any open editors, the error does not appear.